claat/serve.go

```diff
@@ -0,0 +1,221 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"archive/zip"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+)
+
+// cmdServe is the "claat serve ..." subcommand.
+func cmdServe() {
+	var depsDir = "bower_components"
+	var elemDir = "elements"
+	var elemFile = elemDir + "/codelab.html"
+	const codelabElem = `
+<link rel="import" href="../bower_components/google-codelab-elements/google-codelab-elements.html">
+`
+	err := os.MkdirAll(depsDir, 0755)
+	if err != nil {
+		fatalf(err.Error())
+	}
+	// Go get the dependencies.
+	err = fetchRepo(depsDir, "googlecodelabs/codelab-components#2.0.2")
+	if err != nil {
+		fatalf(err.Error())
+	}
+	os.Rename(depsDir+"/codelab-components", depsDir+"/google-codelab-elements")
+	os.Rename(depsDir+"/code-prettify", depsDir+"/google-prettify")
+	err = os.MkdirAll(elemDir, 0755)
+	if err != nil {
+		fatalf(err.Error())
+	}
+	if _, err := os.Stat(elemFile); os.IsNotExist(err) {
+		f, err := os.Create(elemFile)
+		if err != nil {
+			fatalf(err.Error())
+		}
+		defer f.Close()
+		f.WriteString(codelabElem)
+	}
+
+	http.Handle("/", http.FileServer(http.Dir(".")))
+	fmt.Println("Serving on " + *addr + ", opening browser tab now...")
+	openBrowser("http://" + *addr)
+	fatalf("claat serve: %+v", http.ListenAndServe(*addr, nil))
+}
+
+// downloadFile will download a url to a local file. It's efficient because it will
+// write as it downloads and not load the whole file into memory.
+func downloadFile(filepath string, url string) error {
+	fmt.Println("Downloading " + url)
```

Comment:
Remove fmt.Println? "No news is good news."

Reply:
reduced all those prints to one line whenever we need to fetch one or more deps.